### ISSUE TITLE
Do not meet allowed affinity for given DALI thread

### ISSUE THREAD

**Issue by hiyijian** [NVIDIA/DALI#1209]
Dear all,
  I got below warning when setup pipeline and whole process got extremly slow
```
[/opt/dali/dali/util/nvml.h:119] CPU affinity requested by user or recommended by nvml setting does not meet allowed affinity for given DALI thread. Use taskset tool to check allowed affinity
Stacktrace (6 entries):
[frame 0]: /usr/local/lib/python2.7/dist-packages/nvidia/dali/libdali.so(+0xafd5e) [0x7fd92823ed5e]
[frame 1]: /usr/local/lib/python2.7/dist-packages/nvidia/dali/libdali.so(+0x1ce508) [0x7fd92835d508]
[frame 2]: /usr/local/lib/python2.7/dist-packages/nvidia/dali/libdali.so(dali::ThreadPool::ThreadMain(int, int, bool)+0x7c8) [0x7fd928359dd8]
[frame 3]: /usr/local/lib/python2.7/dist-packages/nvidia/dali/libdali.so(+0xe17630) [0x7fd928fa6630]
[frame 4]: /lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba) [0x7fd98776a6ba]
[frame 5]: /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7fd9874a041d]

```

  below is my test scripts "test.py"
```
import random, cv2
import numpy as np
from nvidia.dali.pipeline import Pipeline
import nvidia.dali.ops as ops
import nvidia.dali.types as types

try:
  from mpi4py import MPI
  comm = MPI.COMM_WORLD
  rank = comm.Get_rank()
  size = comm.Get_size()
except:
  rank = 0
  size = 1

class FacePipeline(Pipeline):
  def __init__(self, source, batch_size, shuffle, data_size, mean, scale, mirror, num_threads, device_id):
    super(FacePipeline, self).__init__(batch_size, num_threads, device_id)
    self.mirror = mirror 
    self.input = ops.FileReader(file_root='', file_list=source, random_shuffle=shuffle, 
      num_shards=size, shard_id=rank)
    self.decode = ops.ImageDecoder(device='mixed', output_type=types.BGR)
    self.resize = ops.Resize(device='gpu', image_type=types.BGR, resize_x=data_size, resize_y=data_size)
    self.rng = ops.CoinFlip(probability=0.5)
    self.mirror_normalize = ops.CropMirrorNormalize(device='gpu', mean=mean, std=scale)

  def define_graph(self):
    jpeg, label = self.input()
    img = self.decode(jpeg)
    img = self.resize(img)
    if self.mirror:
      mirror = self.rng()
      img = self.mirror_normalize(img, mirror=mirror)
    else:
      img = self.mirror_normalize(img)
    return (img, label)

pipe = FacePipeline('test.list', 2, True, 112, 127.5, 0.0078125, True, 3, 0)
pipe.build()
for i in range(10):
  pipe_out = pipe.run()
  images, labels = pipe_out
  images = images.as_cpu()
  assert len(images) == len(labels)
  print i
```
  Run this script by below command got the warning
```
mpirun --allow-run-as-root -n 4 python test.py
```
  and below command works fine
```
 python test.py
```

**Comment by Kylin9511** [comment]
I met the same problem...

What confuse me most is: the same DALI dataloader works for apex, but horovod version failed with "Do not meet allowed affinity for given DALI thread"

**Comment by awolant** [comment]
Hello, thanks for reporting the issue.
There seems to be some kind of bug here. We will look into this.

**Comment by hiyijian** [comment]
@Kylin9511 For my case, it gets extremly  slow during pipeline.run(), rather than failure. I put DALI pipleline into python layer of OpenMPI version for caffe.
@awolant In fact, "mpirun -n 4 python test.py" works fine too, even though that warning appears. However,  pipeline.run() gets extremly  slow or even hang if It is putted into python layer of OpenMPI version for caffe.

**Comment by JanuszL** [comment]
Hi,
DALI uses [nvml](https://developer.nvidia.com/nvidia-management-library-nvml) to obtain the recommended CPU affinity for a given GPU. This warning is shown because the CPU affinity mask of the thread returned by `pthread_getaffinity_np` call `and`ed with the set returned with `nvml` is empty.
We need to check why nvml and mpi return completely different values.

**Comment by JanuszL** [comment]
@hiyijian - can you tell me how you "put DALI into python layer of OpenMPI version for caffe.". You mean just `mpirun --allow-run-as-root -n 4 python test.py` ?

**Comment by JanuszL** [comment]
What I want to have a self-contained repro script/scenario. Otherwise, it is very hard to solve this problem.

**Comment by jramapuram** [comment]
I'm running into a similar situation on our slurm cluster. 
The cluster automatically sets cpu affinity based on scheduling, eg: 

```bash
> taskset -p 17843
pid 17843's current affinity mask: 1c0

> taskset -cp 17843
pid 17843's current affinity list: 6-8
```

Do these need to correspond to DALI? I.e. `DALI_AFFINITY_MASK=6,7,8` ?

**Comment by JanuszL** [comment]
> The cluster automatically sets cpu affinity based on scheduling, eg:

If you see this warning it means that the affinity set by the cluster is not the one recommended by the nvml that DALI uses and no affinity is set by DALI on its own. I think you should talk with the cluster administrator regarding the strategy how the affinity is set. As a sanity check, you can run `nvidia-smi topo --matrix` and see if your processes follow the recommended affinity.

**Comment by jramapuram** [comment]
Thanks for the quick response @JanuszL . Just to clarify then, given this matrix:

<img width="775" alt="image" src="https://user-images.githubusercontent.com/8204807/80652392-13bbba00-8a78-11ea-9124-fe522b17c696.png">

does this imply that jobs running on GPU0, GPU1 and GPU2 should be bound with CPU Affinity 0-9 (which I guess means first 10 cores?).  What I'm seeing is that for the following processes:

<img width="561" alt="image" src="https://user-images.githubusercontent.com/8204807/80652451-3cdc4a80-8a78-11ea-9257-50ca5beb03ae.png">

I get the following affinity:

<img width="317" alt="image" src="https://user-images.githubusercontent.com/8204807/80652485-4a91d000-8a78-11ea-96cb-19513cac8fc7.png">

Is it possible to just force DALI to the correct affinity set by the cluster via the environment variable `DALI_AFFINITY_MASK`? And would having a non-matched affinity cause a deadlock like I'm experiencing?

**Comment by JanuszL** [comment]
> does this imply that jobs running on GPU0, GPU1 and GPU2 should be bound with CPU Affinity 0-9

This is the recommended setting based on the system's topology. No following would decrease the performance but should rather not freeze the process pegging the GPU @ 100%. like in your case.
Setting `DALI_AFFINITY_MASK` according to [this document section](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/advanced_topics.html#thread-affinity) should set the proper affinity to worker threads inside DALI, but the main process would still follow what your cluster sets.
You can also check [this script](https://github.com/mlperf/training_results_v0.6/blob/master/NVIDIA/benchmarks/ssd/implementations/pytorch/bind_launch.py) and use `numactl`.

**Comment by jramapuram** [comment]
Thank again, will take a look at that. After further examination of the jobs (from within the jobs as opposed to SSH-ing to the node and checking) it looks like the CPU affinity is being correctly set. I checked the affinity from the job itself via `taskset -cp $$` where `$$` expands to the current PID. 

<img width="775" alt="image" src="https://user-images.githubusercontent.com/8204807/80653869-621e8800-8a7b-11ea-9998-833a75eb34bc.png">

and another example:

<img width="773" alt="image" src="https://user-images.githubusercontent.com/8204807/80653903-719dd100-8a7b-11ea-87a8-f41c6ed4bcde.png">

I went ahead and checked all 13 jobs and all of them correctly had their PID affinity matching the one from the `nvidia-smi topo --matrix` command.
